File: 2024_ruby/aoc/day02.rb
DAY2_INPUT_FILE = "../2024_inputs/day02.txt"
# DAY2_INPUT_FILE = "../2024_inputs/day02_example.txt"

def day2_part1
  reports = File.readlines(DAY2_INPUT_FILE).map { _1.split(/\s+/).map(&:to_i) }

  safe_reports = reports.filter do |levels|
    sorted = levels.sort
    next unless levels == sorted || levels == sorted.reverse

    diffs = levels.each_cons(2).map { (_1 - _2).abs }
    next unless diffs.min >= 1 && diffs.max <= 3

    true
  end

  safe_reports.count
end

File: 2024_rust/src/day02.rs
// const INPUT: &'static str = include_str!("../../2024_inputs/day02.txt");
const INPUT: &'static str = include_str!("../../2024_inputs/day02_example.txt");

pub fn part_one() -> i32 {
    let parse_line = |line: &str| -> Vec<i32> {
        line.split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect()
    };

    INPUT
        .lines()
        .map(parse_line)
        .collect::<Vec<Vec<i32>>>()
        .into_iter()
        .filter(|levels| {
            let ascending = levels.windows(2).all(|pair| pair[0] < pair[1]);
            let descending = levels.windows(2).all(|pair| pair[0] > pair[1]);
            if !ascending && !descending {
                return false;
            }

            let diffs: Vec<i32> = levels
                .windows(2)
                .map(|pair| (pair[0] - pair[1]).abs())
                .collect();

            let min = *diffs.iter().min().unwrap();
            let max = *diffs.iter().max().unwrap();
            if min < 1 || max > 3 {
                return false;
            }

            true
        })
        .count()
        .try_into()
        .unwrap()
}
